### PR TITLE
refactor: add learn.py

### LINKED ISSUE #392: Add learn.py
Refactor the Draco-learn notebook and extract the relevant logic and move it to the the Draco package itself. This makes it more reusable.

### PR DESCRIPTION
**Pull request by JunranY**
close #392

### CODE REVIEW

**Comment by JunranY** [comment]
Added learn.py and data_utils that turns data pairs to training input. 
modified draco_learn notebook accordingly
fixed the math expression in doc

**Problem**: code coverage: I wasn't sure what test cases to write for draco-learn?

File: draco/data_utils.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        violations = draco.count_preferences(spec)
+        if violations is not None:
+            processed_specs[key] = violations
+    return processed_specs[key]
+
+
+def pairs_to_vec(
+    specs: Dict[str, Dict], fields: Tuple[str, str] = ("negative", "positive")
+) -> pd.DataFrame:
+    """given specs, convert them into feature vectors."""
```
Comment (domoritz) [inline draco/data_utils.py:125]: Make this a proper sentence

File: draco/data_utils.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+from .draco import Draco
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+draco = Draco()
+
+
+def get_nested_index(fields=None):
```
Comment (domoritz) [inline draco/data_utils.py:18]: Can you add types?

File: draco/learn.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        clf.fit(X_train, y_train)
+
+        if not quiet:
+            print("Train score: ", clf.score(X_train, y_train))
+            if test_size > 0:
+                print("Dev score: ", clf.score(X_dev, np.ones(len(X_dev))))
+
+        return clf
+
+    def train_and_plot(self, test_size: float = 0.3):
```
Comment (domoritz) [inline draco/learn.py:60]: We should separate the training from the plotting. This way people can use only the training methods which is what most people will want. Moreover, we can then test only the training method. The plotting could either live in a notebook (I think that makes sense given that it's not polished yet) or I a separate method that we mark as not tested.

File: draco/learn.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+from matplotlib.colors import ListedColormap
+from sklearn import svm
+from sklearn.decomposition import PCA
+from sklearn.model_selection import train_test_split
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class DracoLearn:
```
Comment (domoritz) [inline draco/learn.py:15]: I don't know whether this needs to be a class vs just a collection of methods. Is there benefit to having a class?

**Review by domoritz — CHANGES_REQUESTED** [review]
Remove

File: draco/data_utils.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        if violations is not None:
+            processed_specs[key] = violations
+    return processed_specs[key]
+
+
+def pairs_to_vec(
+    specs: Dict[str, Dict], fields: Tuple[str, str] = ("negative", "positive")
+) -> pd.DataFrame:
+    """
+    Given pairs of positive and negative draco specification, count the number of times
```
Comment (domoritz) [inline draco/data_utils.py:131]:
```suggestion
    Given pairs of positive and negative Draco specifications, count the number of times
```

**Review by domoritz — APPROVED** [review]

**Pull request merged**